src/App.jsx

```diff
@@ -1,46 +1,245 @@
+          }
+
+          return todo;
+        });
+
+        setTodoList(fetchedTodos);
+      } catch (error) {
+        setErrorMessage(error.message);
+        console.error(error);
+      } finally {
+        setIsLoading(false);
+      }
+    };
+
+    fetchTodos();
+  }, []);
+
+  const handleAddTodo = async (newTodo) => {
+    const payload = {
+      records: [
+        {
+          fields: {
+            title: newTodo.title,
+            isCompleted: false,
+          },
+        },
+      ],
+    };
+
+    const options = {
+      method: 'POST',
+      headers: {
+        Authorization: token,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(payload),
+    };
+
+    try {
+      setIsSaving(true);
+      const resp = await fetch(url, options);
+
+      if (!resp.ok) {
+        throw new Error(resp.statusText || 'Failed to save todo');
+      }
+
+      const { records } = await resp.json();
+      const savedTodo = {
+        id: records[0].id,
+        ...records[0].fields,
+      };
+
+      if (!savedTodo.isCompleted) {
+        savedTodo.isCompleted = false;
+      }
+
+      setTodoList([...todoList, savedTodo]);
+    } catch (error) {
+      console.error(error);
+      setErrorMessage(error.message);
+    } finally {
+      setIsSaving(false);
+    }
   };
 
-  const completeTodo = (id) => {
-    const updatedTodos = todoList.map((todo) => {
+  const completeTodo = async (id) => {
+    const todoToComplete = todoList.find((todo) => todo.id === id);
+    const originalTodo = { ...todoToComplete };
+
+    const optimisticTodos = todoList.map((todo) => {
       if (todo.id === id) {
         return { ...todo, isCompleted: true };
       }
       return todo;
     });
-    setTodoList(updatedTodos);
+
+    setTodoList(optimisticTodos);
+
+    const payload = {
+      records: [
+        {
+          id: id,
+          fields: {
+            title: todoToComplete.title,
+            isCompleted: true,
+          },
+        },
+      ],
+    };
+
+    const options = {
+      method: 'PATCH',
+      headers: {
+        Authorization: token,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(payload),
+    };
+
+    try {
+      const resp = await fetch(url, options);
+
+      if (!resp.ok) {
+        throw new Error(resp.statusText || 'Failed to complete todo');
+      }
+    } catch (error) {
+      console.error(error);
+      setErrorMessage(`${error.message}. Reverting todo...`);
+
+      const revertedTodos = todoList.map((todo) => {
+        if (todo.id === id) {
+          return originalTodo;
+        }
+        return todo;
+      });
+
+      setTodoList(revertedTodos);
+    }
   };
 
-  const updateTodo = (editedTodo) => {
-    const updatedTodos = todoList.map(todo => {
+  const updateTodo = async (editedTodo) => {
+    const originalTodo = todoList.find((todo) => todo.id === editedTodo.id);
+
+    const optimisticTodos = todoList.map((todo) => {
```

Comment:
Good instinct in trying to give a descriptive variable name but `optimisticTodos` doesn't really add much context, it's actually a bit misleading and confusing. Since "optimistic" is a system used to describe what a function is doing it would better server as a function name. e.g. `optimisticUpdateTodos`

Reply:
This applies to other places you've used "optimistic" .